docs: Clarify Session Modes and improve other docs (#116)
getsentry/develop: src/docs/sdk/sessions.mdx

### Session Counting / init

It's crucial that the initial session update sent to the system has `init` set
to `true`. This is necessary because the server currently does not

> deduplicate the total session count as an optimization.  If the initial

Reviewer: ```suggestion
deduplicate the total session count as an optimization. If the initial
```

Author: :-D you don’t like these double-spaces? Neither does my auto-formatter btw, so I just reformatted everything.

### Exited

A session can transition to `exited` which is exactly the same state as `ok`
with one difference: sessions transitioned to `exited` have their session
duration averaged. This lets Sentry show you the duration of non crashed
sessions over time.

## SDK Considerations

Generally speaking there are two separate modes for health reporting that SDKs
can use. One is very short lived sessions, the other are user attended sessions.

**Short lived sessions** _(server-mode / request-mode)_

These sessions roughly correspond to HTTP requests or RPC calls in a
server setting.

- high in volume, typically one session for each request
- the number of sessions is usually higher than the number of sentry events
- sessions are attached to a single hub / concurrency unit
- timing information is typically useless because session time in the milliseconds

**User attended sessions** _(user-mode / application-mode)_

These are sessions that are more corresponding to an actual user session or
application run. This is what you would see in a web browser, mobile world,
command line application or similar.

- typically just a single session from application start to quit

> - there are usually fewer sessions than sentry events

Reviewer: ```suggestion
- there are usually fewer sessions than Sentry events
```

- sessions span multiple hubs / threads
- session duration typically in the minutes, timing information is useful

Both of those cases look similar from the API point of view but different
recommendations apply for SDKs.

### Choosing the Session Mode

While it is in theory possible to use both session modes in a single application,
it is recommended that the SDK default to a single mode that is most

> appropriate to the main use-case of the language ecosystem. This is similar to,

Reviewer: ```suggestion
appropriate to the main use case of the language ecosystem. This is similar to,
```

and can be used in the same way as a _global Hub mode_ that certain SDKs support.

When the SDK is configured to use _user-mode_ sessions, or global Hub mode, a
single session should be started at the start of the application and should
persist through the application's runtime. Depending on the SDK internals, this
single session can be shared among all application threads and thread-local Hubs.

When using _server-mode_ sessions, no application-wide session will be started,

> and it is up to integrations or the user to start the session when the request is received and end it when a response is returned.

Reviewer: ```suggestion
and it is up to integrations or the user to start the session when the request is
received and end it when a response is returned.
```

For consistency with the rest of the paragraph that has wrapped lines.

for instance it makes sense to transition a session to abnormal if it was stored

> but the exit of the application was not observed but also did not crash.  For

Reviewer: ```suggestion
but the exit of the application was not observed but also did not crash. For
```

server setting.

- high in volume, typically one session for each request

> - the number of sessions is usually higher than the number of captured events

Reviewer 2: Sorry - not trying to beat a dead horse here, but captured _session_ events or captured sentry events in general? 

I guess I'm still not clear why this information is pertinent. Is what we're trying to say essentially, "you probably won't capture every request as a session and will probably only send one session event per session" and below, "you'll probably capture every session and each session will likely have multiple session events sent?"

Author: with `events`, I always mean sentry events, otherwise I would call them "session updates".

This is basically to highlight that you want to create a session for every request (maybe even 10k req/s), but not every request will raise an error, and thus a `capture_event`.

command line application or similar.

- typically just a single session from application start to quit
- there are usually fewer sessions than captured events
- sessions span multiple hubs / threads
- session duration typically in the minutes, timing information is useful

Both of those cases look similar from the API point of view but different
recommendations apply for SDKs.

### Choosing the Session Mode

While it is in theory possible to use both session modes in a single application,

> it is recommended that the SDK defaults to a single mode that is most

Reviewer 2: ```suggestion
it is recommended that the SDK default to a single mode that is most
```

appropriate to the main use-case of the language ecosystem. This is similar to,
and can be used in the same way as a *global Hub mode* that certain SDKs support.

When the SDK is configured to use *user-mode* sessions, or global Hub mode, a
single session should be started at the start of the application and should

> persist through the applications runtime. Depending on the SDK internals, this

Reviewer 2: ```suggestion
persist through the application's runtime. Depending on the SDK internals, this
```

single session can be shared among all application threads and thread-local Hubs.

When using *server-mode* sessions, no application-wide session will be started,

> and it is up to integrations or the user to start and end the session for the

Reviewer 2: ```suggestion
and it is up to integrations or the user to start the session when the request is received and end it when a response is returned.
```

(Doing this all in one line (and getting rid of the line below) because it's too complicated otherwise. Really wish we didn't do manual line breaks in our docs...)

> duration of a request.

Reviewer 2: ```suggestion
```

`Hub.start_session()`

> > Stores a session on the current scope and starts tracking it. This normally attaches a brand new `session` to the scope,

Reviewer 2: Are the backticks on the second instance (but not first or third instances) of the word "session" there to imply that the scope ends up with a property called `.session`? Also: "normally" - are there cases where this doesn't happen?
